Source/santad/DataLayer/SNTRuleTable.mm

```diff
@@ -589,4 +591,24 @@ - (BOOL)fillError:(NSError **)error code:(SNTRuleTableError)code message:(NSStri
   return rules;
 }
 
+- (NSString *)hashOfHashes {
+  __block CC_SHA256_CTX sha;
+  CC_SHA256_Init(&sha);
+
+  [self inDatabase:^(FMDatabase *db) {
+    FMResultSet *rs = [db executeQuery:@"SELECT * FROM rules"];
```

Comment:
You might need to exclude transitive rules here as it'll change your hash of hashes and the server will have no idea that these exist.

Reply:
Also any rules added manually while a sync server didn't exist (though I don't think we track such thing today). However this is likely niche...

Reply:
Another thought - there's been requests and talks in the past about per-user rules. It would be good to have a bit of an idea about how something like those might be represented here if they're ever implemented.

Reply:
@pmarkowsky Good call, I've excluded transitive rules

@mlw If a sync server is added the first sync afterward is a clean sync so the hashes shouldn't matter. For per-user rules those would be implemented on the server side anyway, so that _shouldn't_ matter here. The purpose of this is not necessarily to have the server calculate the hash and compare them but to store the hash sent in a postflight and then validate that the one sent in a preflight matches the stored value.

Reply:
@mlw I'm assuming you're talking about the rumblings in the past that folks would like to block execution by the user e.g. root can run X but not user peterm. 

That'd have to be implemented in the client but is an extension to rules similar to #1200. 

Re: Hash of Hashes. Do you need to sort by identifier and policy to create a stable hash? Like if I get a rule out of order for some reason i'll end up with a different hash even though the set of policies is the same.

Reply:
> Do you need to sort by identifier and policy to create a stable hash? Like if I get a rule out of order for some reason i'll end up with a different hash even though the set of policies is the same.

Right now this hashes based on insertion order, which is stable on the client at least. The server could ensure this is stable by always ordering by creation timestamp but I'm not really expecting servers to calculate this in the same way as the client but instead just check the hash doesn't change during preflight.

Reply:
>Right now this hashes based on insertion order, which is stable on the client at least. The server could ensure this is stable by always ordering by creation timestamp but I'm not really expecting servers to calculate this in the same way as the client but instead just check the hash doesn't change during preflight.

That's good to know.  

Not saying you should do this as the hash is probably the right tradeoff for now, but previously, I looked into using an xorfilter/bloomfilter rather than a hash to avoid worrying about ordering e.g. this single header one https://github.com/FastFilter/xor_singleheader.

The advantage to this is that you're creating a set membership data structure so order becomes irrelevant. 

The process would just be:

1. Hash rules (policy, state, identifier) to a 64-bit integer (e.g. use murmur3)
2. Insert into the xorfilter
3. Serialize and ship the xorfilter to the server in postflight
4. Rehydrate the xorfilter on the server and then for all rules per user check if they're in the supplied xorfilter

Alternatively you could also receive a xorfilter from the server of all the expected rules and as you're scanning the Santa rule db check if those rules are supposed to be in the client DB. This would let you know exactly which rules were spurious.

Reply:
I see 3 problems with that approach

1. Serializing and de-serializing the filter may require reimplementing the serialization - the Go implementation in particular doesn't support this
2. This sounds far more expensive on the server-side; what I've proposed requires saving a single extra field in postflight and doing a simple 64-byte string comparison during preflight. This would require looping over all rules in postflight to check that they're in the filter.
3. This will not detect extra rules that have been added. Reversing the process _would_ find extra rules (which is better, imo) but then not detect missing rules.

Reply:
@mlw Gotcha. I don't think that should be an issue, whether we implement that as an extra field in a rule or via CEL, we would add the new field(s) to those included in each rules' hash.